apps/src/javalab/redux/editorRedux.ts

```diff
@@ -0,0 +1,320 @@
+      ) {
+        state.sources[action.payload.filename].isValidation =
+          action.payload.isValidation;
+      },
+      prepare(filename: string, isValidation: boolean) {
+        return {payload: {filename, isValidation}};
+      }
+    },
+    sourceFileOrderUpdated(state) {
+      const sources = _.cloneDeep(state.sources);
+      const updatedSources = updateAllSourceFileOrders(
+        sources,
+        state.fileMetadata,
+        state.orderedTabKeys
+      );
+      state.sources = updatedSources;
+    },
+    removeFile(state, action: PayloadAction<string>) {
+      const filename = action.payload;
+      const newSources = {...state.sources};
+      delete newSources[filename];
+      state.sources = newSources;
+    },
+    openEditorDialog(state, action: PayloadAction<JavlabEditorDialogOptions>) {
+      if (Object.values(JavalabEditorDialog).includes(action.payload)) {
+        state.editorOpenDialogName = action.payload;
+      }
+    },
+    closeEditorDialog(state) {
+      state.editorOpenDialogName = null;
+    },
+    setEditTabKey(state, action: PayloadAction<string>) {
+      state.editTabKey = action.payload;
+    },
+    setActiveTabKey(state, action: PayloadAction<string>) {
+      state.activeTabKey = action.payload;
+    },
+    setFileMetadata(state, action: PayloadAction<FileMetadata>) {
+      state.fileMetadata = action.payload;
+    },
+    setOrderedTabKeys(state, action: PayloadAction<string[]>) {
+      state.orderedTabKeys = action.payload;
+    },
+    setAllEditorMetadata: {
+      reducer(
+        state,
+        action: PayloadAction<{
+          fileMetadata: FileMetadata;
+          orderedTabKeys: string[];
+          activeTabKey: string;
+          lastTabKeyIndex: number;
+        }>
+      ) {
+        state.fileMetadata = action.payload.fileMetadata;
+        state.orderedTabKeys = action.payload.orderedTabKeys;
+        state.activeTabKey = action.payload.activeTabKey;
+        state.lastTabKeyIndex =
+          action.payload.lastTabKeyIndex || state.lastTabKeyIndex;
+      },
+      prepare(
+        fileMetadata: FileMetadata,
+        orderedTabKeys: string[],
+        activeTabKey: string,
+        lastTabKeyIndex: number
+      ) {
+        return {
+          payload: {fileMetadata, orderedTabKeys, activeTabKey, lastTabKeyIndex}
+        };
+      }
+    },
+    setNewFileError(state, action: PayloadAction<string>) {
+      state.newFileError = action.payload;
+    },
+    clearNewFileError(state) {
+      state.newFileError = null;
+    },
+    setRenameFileError(state, action: PayloadAction<string>) {
+      state.renameFileError = action.payload;
+    },
+    clearRenameFileError(state) {
+      state.renameFileError = null;
+    }
+  }
+});
+
+// Selectors
+// We don't have a type for the entire redux store yet, so we need to use any here.
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const selectSources = (state: any) => state.javalabEditor.sources;
```

Comment:
nit: I think you could potentially type this `(state: { javalabEditor: JavalabEditorState })` to avoid the `any`?

Reply:
Oh cool!